### PR TITLE
implement snippet deduplication using MMR

### PR DESCRIPTION
**Pull request by nerdypepper**
- deduplication is on par with `main`
- `main` refuses to collect more than 4 snippets from the same file, MMR does not have an artificial limit.

### CODE REVIEW

File: server/bleep/src/webserver/answer.rs [hunk 1]
```diff
@@ -1092,3 +1085,52 @@ impl StopWatch {
         duration
     }
 }
+
+fn dot(a: &[f32], b: &[f32]) -> f32 {
```
Comment (ggordonhall) [inline server/bleep/src/webserver/answer.rs:1089]: Can you move these functions into `semantic.rs`. They are not specific to the `answer` pipeline and could be applied to any set of semantic search results.

File: server/bleep/src/webserver/answer.rs [hunk 2]
```diff
@@ -82,6 +82,9 @@ pub struct Snippet {
     pub start_byte: usize,
     pub end_byte: usize,
     pub score: f32,
+
+    #[serde(skip)]
+    pub vector: Vec<f32>,
```
Comment (ggordonhall) [inline server/bleep/src/webserver/answer.rs:87]: Can we rename this `embedding` and add a comment explaining where it is used?

File: server/bleep/src/webserver/answer.rs [hunk 3]
```diff
@@ -234,49 +248,23 @@ async fn search_snippets(
...
-                    .or_insert_with(Vec::new)
-                    .push(std::ops::Range {
-                        start: snippet.start_line,
-                        end: snippet.end_line,
-                    });
-                snippets.push(snippet);
-            }
-        }
+fn deduplicate_snippets(all_snippets: Vec<Snippet>, query_embedding: Vec<f32>) -> Vec<Snippet> {
+    let lambda = 0.5;
```
Comment (ggordonhall) [inline server/bleep/src/webserver/answer.rs:260]: Would be nice to have a comment explaining what `lambda` does.

File: server/bleep/src/webserver/answer.rs [hunk 4]
```diff
@@ -442,7 +438,16 @@ async fn handle_inner(
...
                 );
 
-                let filtered_snippets = deduplicate_snippets(all_snippets);
+                let query_embedding = semantic.embed(rephrased_query).map_err(|e| {
+                    error!("failed to embed query: {}", e);
+                    Error::internal(e)
+                })?;
+                let filtered_snippets = deduplicate_snippets(all_snippets, query_embedding);
+
+                event.write().await.stages.push(
```
Comment (nerdypepper) [inline server/bleep/src/webserver/answer.rs:447]: moved the analytics event up here, so we can see if `filtered_snippets` is empty in the pipeline.

**Review by ggordonhall — APPROVED** [review]

**Pull request merged**